keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -0,0 +1,256 @@
+details on post GA tasks.
+
+### v1 API
+
+The CRD `v1` API will be the same as the `v1beta1` but with all changes to the API from the GA tasks:
+
+* Rename misnamed json field [JSONPath](https://github.com/kubernetes/kubernetes/blob/06bc7e3e0026ea25065f59f4bd305c0b7dbbc145/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L226-L227) to `jsonPath`
+* [Replace top-level fields with per version fields](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresource-conversion-webhook.md#top-level-fields-to-per-version-fields)
+* Restrict OpenAPI per [Vanilla OpenAPI Subset Design](https://docs.google.com/document/d/1pcGlbmw-2Y0JJs9hsYnSBXamgG9TfWtHY6eh80zSTd8)
+
+## Test Plan
+
+### Integration tests for GA
+
+TODO: complete this list
+
+* [Pruning for CustomResources KEP](https://github.com/kubernetes/enhancements/pull/709)
+  * TODO
+* [Vanilla OpenAPI Subset Design](https://docs.google.com/document/d/1pcGlbmw-2Y0JJs9hsYnSBXamgG9TfWtHY6eh80zSTd8)
+  * TODO
+* CRD versioning/conversion (https://github.com/kubernetes/kubernetes/issues/64136):
+  * Ensure what is persisted in etcd matches the storage version
+  * Set up a CRD, persist some data, changed the version, and ensure the previously persisted data is readable
+  * Ensure discovery docs track a CRD through creation, version addition, version removal, and deletion
+  * Ensure `spec.served` is respected
+
+### e2e tests for GA
+
+TODO: complete this list
+
+* Ensure v1beta1 and v1 CRDs are round trippable
+* Ensure CR watch observes events on create/update/delete
+* Ensure a CR watch is correctly terminated if CRD is deleted
+
+### Conformance plan for GA
+
+TODO: complete this list
+
+* Promote existing e2e tests:
+  * [CustomResourceDefinition Watch](https://github.com/kubernetes/kubernetes/blob/36ccff1b27b6aee5c5599334842cb2b8cf8063f3/test/e2e/apimachinery/crd_watch.go#L37)
+  * [CustomResourcePublishOpenAPI](https://github.com/kubernetes/kubernetes/blob/36ccff1b27b6aee5c5599334842cb2b8cf8063f3/test/e2e/apimachinery/crd_publish_openapi.go#L52)
+  * [CustomResourceConversionWebhook](https://github.com/kubernetes/kubernetes/blob/36ccff1b27b6aee5c5599334842cb2b8cf8063f3/test/e2e/apimachinery/crd_conversion_webhook.go#L81)
+
+### Scale Targets for GA
+
+TODO: quantify our scale targets
+
+* CRD resource count scales to X within SLO (watches, QPS, writes/s)
+* Webhook conversion QPS
+
+## Graduation Criteria
+
+To mark these as complete, all of the above features need to be implemented. An
+umbrella issue is tracking all of these changes. Also there need to be
+sufficient tests for any of these new features and all existing features and
+documentation should be completed for all features.
+
+See [umbrella issue](https://github.com/kubernetes/kubernetes/issues/58682) for status.
+
+## Post-GA tasks
```

Comment:
"Potential Post-GA tasks"

Some of them below are far from consensus or good enough understanding. We won't decide anything about those in this KEP.

Reply:
Maybe we should not try to list them comprehensibly at all or give detail below. A link to an umbrella github issue is enough IMO.

Reply:
Sounds good. I've replaced this section with a link to the umbrella issue where readers can find an up-to-date post GA task list.

Reply:
Re-added this per @lavalamp's request. Added a note that the list is a snapshot from when KEP was written and to see the umbrella issue for current status.

Reply:
looks good.

Reply:
Yeah, the umbrella issue is not a good historical record. The purpose of this section is to inform people that we didn't forget about these things, we just decided they don't block v1 GA--you might want to make that a bit clearer in the text in a followup.